boa/src/value/serde_json.rs

```diff
@@ -0,0 +1,157 @@
+//! This module implements the conversions from and into [`serde_json::Value`].
+
+use super::JsValue;
+use crate::{
+    builtins::Array,
+    property::{PropertyDescriptor, PropertyKey},
+    Context, JsResult,
+};
+use serde_json::{Map, Value};
+
+impl JsValue {
+    /// Converts a [`serde_json::Value`] to a `JsValue`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use boa::{Context, JsValue};
+    ///
+    /// let data = r#"
+    ///     {
+    ///         "name": "John Doe",
+    ///         "age": 43,
+    ///         "phones": [
+    ///             "+44 1234567",
+    ///             "+44 2345678"
+    ///         ]
+    ///      }"#;
+    ///
+    /// let json: serde_json::Value = serde_json::from_str(data).unwrap();
+    ///
+    /// let mut context = Context::default();
+    /// let value = JsValue::from_json(&json, &mut context).unwrap();
+    /// #
+    /// # assert_eq!(json, value.to_json(&mut context).unwrap());
+    /// ```
+    pub fn from_json(json: &Value, context: &mut Context) -> JsResult<Self> {
+        /// Biggest possible integer, as i64.
+        const MAX_INT: i64 = i32::MAX as i64;
+
+        match json {
+            Value::Null => Ok(Self::Null),
+            Value::Bool(b) => Ok(Self::Boolean(*b)),
+            Value::Number(num) => match num.as_i64() {
+                Some(s @ 0..=MAX_INT) => Ok(Self::Integer(s as i32)),
+                Some(s) => Ok(Self::Rational(s as f64)),
+                None => {
+                    if let Some(i) = num.as_u64() {
+                        Ok(Self::Rational(i as f64))
+                    } else {
+                        Ok(Self::Rational(num.as_f64().ok_or_else(|| {
+                            context.construct_type_error(format!(
+                                "could not convert JSON number {num} to JsValue"
+                            ))
+                        })?))
```

Comment:
A more functional approach should be more readable, I think.

Something like:
`num.as_i64().filter({range}).map(|i|JsValue::Integer(i as i32)).or_else(||num.as_f64()).map(JsValue::Rational).ok_or_else({error})?`